### PR TITLE
GPU Treeshap in FIL Backend

### PR DESCRIPTION
**Pull request by divyegala**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -103,9 +110,9 @@ else()
 
   if(TRITON_ENABLE_GPU)
     rapids_cuda_init_architectures(RAPIDS_TRITON_BACKEND)
-    project(RAPIDS_TRITON_BACKEND VERSION 21.12.00 LANGUAGES CXX CUDA)
+    project(RAPIDS_TRITON_BACKEND VERSION 22.02.00 LANGUAGES CXX CUDA)
```
Comment (hcho3) [inline CMakeLists.txt:113]: We need to update this to `22.04` (and a few other places).

Comment (divyegala) [reply CMakeLists.txt:113]: This is the version of the FIL backend, which is going to be updated to `22.03` this release

Comment (hcho3) [reply CMakeLists.txt:113]: Actually never mind. Didn't realize Rapids-Triton was following its own release cycle.

File: cmake/thirdparty/get_cuml.cmake [hunk 2]
```diff
@@ -37,7 +56,8 @@ endfunction()
 # Change pinned tag here to test a commit in CI
 # To use a different RAFT locally, set the CMake variable
 # CPM_raft_SOURCE=/path/to/local/raft
-find_and_configure_cuml(VERSION    21.12
-                        FORK       rapidsai
-                        PINNED_TAG branch-21.12
+find_and_configure_cuml(VERSION    ${RAPIDS_TRITON_MIN_VERSION_rapids_projects}
+                        FORK       dantegd
+                        PINNED_TAG 2112-libcuml-configurable
```
Comment (hcho3) [inline cmake/thirdparty/get_cuml.cmake:61]: This will be changed back to `branch-22.04` once https://github.com/rapidsai/cuml/pull/4296 is merged?

Comment (divyegala) [reply cmake/thirdparty/get_cuml.cmake:61]: We might want to pin the actual commit hash for that merge.

File: CMakeLists.txt [hunk 3]
```diff
@@ -67,7 +67,7 @@ else()
 
   ##############################################################################
   # - Prepare rapids-cmake -----------------------------------------------------
-  file(DOWNLOAD https://raw.githubusercontent.com/rapidsai/rapids-cmake/branch-21.12/RAPIDS.cmake
+  file(DOWNLOAD https://raw.githubusercontent.com/rapidsai/rapids-cmake/branch-22.04/RAPIDS.cmake
```
Comment (dantegd) [inline CMakeLists.txt:70]: I wonder if we should pin a particular tag for rap-cmake

File: CMakeLists.txt [hunk 4]
```diff
@@ -86,12 +86,19 @@ else()
...
+
 
   message(VERBOSE "RAPIDS_TRITON_BACKEND: Enabling detection of conda environment for dependencies: ${DETECT_CONDA_ENV}")
   message(VERBOSE "RAPIDS_TRITON_BACKEND: Enabling kernelinfo in nvcc: ${CUDA_ENABLE_KERNEL_INFO}")
   message(VERBOSE "RAPIDS_TRITON_BACKEND: Enabling lineinfo in nvcc: ${CUDA_ENABLE_LINE_INFO}")
   message(VERBOSE "RAPIDS_TRITON_BACKEND: Enabling nvtx markers: ${NVTX}")
   message(VERBOSE "RAPIDS_TRITON_BACKEND: Build RAPIDS_TRITON_BACKEND unit-tests: ${BUILD_TESTS}")
+  message(VERBOSE "RAPIDS_TRITON_BACKEND: RAPIDS projects dependencies version: ${RAPIDS_DEPENDENCIES_VERSION}")
+  message(VERBOSE "RAPIDS_TRITON_BACKEND: Linking Treelite statically: ${FIL_TRITON_USE_TREELITE_STATIC}")
+  message(VERBOSE "RAPIDS_TRITON_BACKEND: Compiling GPU Treeshap in cuml++.so ${TRITON_FIL_ENABLE_TREESHAP}")
```
Comment (dantegd) [inline CMakeLists.txt:101]:
```suggestion
  message(VERBOSE "RAPIDS_TRITON_BACKEND: Compiling GPU Treeshap in libcuml++.so ${TRITON_FIL_ENABLE_TREESHAP}")
```

Verbosity to help newcomers

File: CMakeLists.txt (hunk 1 above)
Comment (dantegd) [inline CMakeLists.txt:113]:
```suggestion
    project(RAPIDS_TRITON_BACKEND VERSION 22.03.00 LANGUAGES CXX CUDA)
```

shouldn't this be 22.03?

File: CMakeLists.txt [hunk 5]
```diff
@@ -103,9 +110,9 @@ else()
 
   if(TRITON_ENABLE_GPU)
     rapids_cuda_init_architectures(RAPIDS_TRITON_BACKEND)
-    project(RAPIDS_TRITON_BACKEND VERSION 21.12.00 LANGUAGES CXX CUDA)
+    project(RAPIDS_TRITON_BACKEND VERSION 22.02.00 LANGUAGES CXX CUDA)
   else()
-    project(RAPIDS_TRITON_BACKEND VERSION 21.12.00 LANGUAGES CXX)
+    project(RAPIDS_TRITON_BACKEND VERSION 22.02.00 LANGUAGES CXX)
```
Comment (dantegd) [inline CMakeLists.txt:115]:
```suggestion
    project(RAPIDS_TRITON_BACKEND VERSION 22.03.00 LANGUAGES CXX)
```

Same as above

File: CMakeLists.txt [hunk 6]
```diff
@@ -205,44 +216,45 @@ else()
...
+        TRITON_FIL_ENABLE_TREESHAP=1
     )
   endif()
   
-  if(TRITON_ENABLE_GPU)
-    install(
-      TARGETS ${BACKEND_TARGET} cuml++
-      LIBRARY DESTINATION ${BACKEND_FOLDER}/${BACKEND_NAME}
-    )
+  if(NOT TRITON_FIL_USE_TREELITE_STATIC)
```
Comment (dantegd) [inline CMakeLists.txt:227]: I know we do this pattern in cuml right now, but it'd be a little cleaner to not do a negative conditional, instead do `if(NOT TRITON_FIL_USE_TREELITE_STATIC)`, slightly easier to read

File: CMakeLists.txt [hunk 7]
```diff
@@ -205,44 +216,45 @@ else()
...
   endif()
+
+  target_link_libraries(${BACKEND_TARGET}
+  PRIVATE
+    $<$<BOOL:${TRITON_ENABLE_GPU}>:cuml++>
+    ${TREELITE_LIBS}
+    rapids_triton::rapids_triton
+    triton-core-serverstub
+    triton-backend-utils
+    "${TRITONSERVER_LIB}"
```
Comment (dantegd) [inline CMakeLists.txt:242]: Unrelated to changes in the PR so we can address it later, but are this quotes necessary?

File: cmake/thirdparty/get_cuml.cmake (hunk 2 above)
Comment (dantegd) [reply cmake/thirdparty/get_cuml.cmake:61]: More precisely, we are planning to create a particular branch to pin RMM, RAFT, cuML and RAPIDS-Triton to be entirely fixed for release 22.03 tomorrow or Friday

File: cmake/thirdparty/get_rapids-triton.cmake [hunk 8]
```diff
@@ -40,7 +40,7 @@ endfunction()
 # Change pinned tag here to test a commit in CI
 # To use a different RAFT locally, set the CMake variable
 # CPM_raft_SOURCE=/path/to/local/raft
-find_and_configure_rapids_triton(VERSION    21.12
-                        FORK       rapidsai
-                        PINNED_TAG branch-21.12
-                        )
+find_and_configure_rapids_triton(VERSION    ${RAPIDS_TRITON_MIN_VERSION_rapids_projects}
+                                 FORK       divyegala
```
Comment (dantegd) [inline cmake/thirdparty/get_rapids-triton.cmake:44]: Reminder to change this back when the merge is done in RT

**Review by dantegd — CHANGES_REQUESTED** [review]
Nothing major needed on the cmake/docker side, just minor things

File: CMakeLists.txt (hunk 3 above)
Comment (divyegala) [reply CMakeLists.txt:70]: The one for tomorrow or Friday? Going by your comment above for specific branches

File: CMakeLists.txt (hunk 7 above)
Comment (divyegala) [reply CMakeLists.txt:242]: Not quite sure, I believe it's an environment variable and not a cmake variable. Possibly the reason? cc @wphicks

Comment (wphicks) [reply CMakeLists.txt:242]: I honestly don't remember. If I have a moment later, I'll review and see if I can understand what I was thinking.

File: src/gpu_treeshap_model.h [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  TreeShapModel(TreeShapModel&& other) = default;
+  TreeShapModel& operator=(TreeShapModel&& other) = default;
+
+  void predict(
+      rapids::Buffer<float>& output, rapids::Buffer<float const> const& input,
+      std::size_t n_rows, std::size_t n_cols) const
+  {
+    // Need to synchronize on the stream because treeshap currently does not
+    // take a stream on its API
+    raft_handle_.sync_stream();
```
Comment (wphicks) [inline src/gpu_treeshap_model.h:54]: We may actually prefer
```
    input.stream_synchronize();
    output.stream_synchronize();
```
In theory, a later refactor could set up the input and output buffers to synchronize on separate streams from the model itself for performance reasons.

File: src/model.h [hunk 10]
```diff
@@ -110,6 +116,50 @@ struct RapidsModel : rapids::Model<RapidsSharedState> {
...
+        auto treeshap_output_buffer = rapids::Buffer<float>(
+            treeshap_output.data(), treeshap_output.size(),
+            treeshap_output.mem_type(), treeshap_output.device(),
+            treeshap_output.stream());
+
+        if (gpu_treeshap_model.has_value()) {
+          // Always copy input buffer to device memory
+          if (input.mem_type() == rapids::HostMemory &&
+              input_buffer.mem_type() == rapids::HostMemory) {
+            input_buffer = rapids::Buffer<float const>(
```
Comment (wphicks) [inline src/model.h:139]: Could you add a `TODO` just saying that we should look at collapsing this `input_buffer` with the one used in ordinary inference? The logic is a little tricky, so I don't want us to do it right now, but doing so would save us a host-to-device copy.

File: src/model.h [hunk 11]
```diff
@@ -110,6 +116,50 @@ struct RapidsModel : rapids::Model<RapidsSharedState> {
...
+              // Create output buffer in correct  location
+              treeshap_output_buffer = rapids::Buffer<float>(
+                  treeshap_output.size(), input_buffer.mem_type(),
+                  get_device_id(), get_stream());
+            }
+          }
+        }
+
+        // The shape of treeshap output is (, num_classes * (n_cols + 1))
+        gpu_treeshap_model.value().predict(
```
Comment (wphicks) [inline src/model.h:153]: Let's move this into the `if (gpu_treeshap_model.has_value())` branch and then avoid the error checking via:
```suggestion
        gpu_treeshap_model.value()->predict(
```

**Review by wphicks — CHANGES_REQUESTED** [review]

File: conda/environments/triton_test.yml [hunk 12]
```diff
@@ -3,21 +3,22 @@ channels:
...
+  - cuml=22.04
   - flake8
   - hypothesis
   - lightgbm
   - matplotlib
   - pip
   - pytest
   - python=3.8
-  - rapidsai::xgboost
+  - rapidsai-nightly::xgboost
```
[the remark this answers is not in the document]
Comment (divyegala) [reply conda/environments/triton_test.yml:19]: We weren't able to solve the environment with stable, this is why this change was needed

File: qa/L0_e2e/test_model.py [hunk 13]
```diff
@@ -12,6 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from ast import mod
```
[the remark this answers is not in the document]
Comment (divyegala) [reply qa/L0_e2e/test_model.py:15]: I don't think I added this. @hcho3 is this from your PR?

File: src/gpu_treeshap_model.h [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+
+template <>
+struct TreeShapModel<rapids::DeviceMemory> {
+  using device_id_t = int;
+  TreeShapModel(
+      device_id_t device_id, cudaStream_t stream,
+      std::shared_ptr<TreeliteModel> tl_model)
+      : device_id_{device_id}, raft_handle_{stream}, tl_model_{tl_model}
+  {
+    path_info_ = ML::Explainer::extract_path_info(tl_model_->handle());
```
[the remark this answers is not in the document]
Comment (divyegala) [reply src/gpu_treeshap_model.h:40]: None, I just had one of those moments..

File: src/model.h [hunk 15]
```diff
@@ -110,6 +116,50 @@ struct RapidsModel : rapids::Model<RapidsSharedState> {
...
+      // boolean to check whether gpu treeshap will be executed
+      auto run_gpu_treeshap =
+          shared_state->check_output_name("treeshap_output");
+
+      if (run_gpu_treeshap) {
+        auto treeshap_output = get_output<float>(batch, "treeshap_output");
+        auto treeshap_output_buffer = rapids::Buffer<float>(
+            treeshap_output.data(), treeshap_output.size(),
+            treeshap_output.mem_type(), treeshap_output.device(),
+            treeshap_output.stream());
```
[the remark this answers is not in the document]
Comment (divyegala) [reply src/model.h:133]: Shall I refactor the example above as well for the normal `output_buffer`?

File: qa/L0_e2e/test_model.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (hcho3) [reply qa/L0_e2e/test_model.py:15]: I don't recall adding it.

File: src/model.h (hunk 10 above)
Comment (divyegala) [reply src/model.h:139]: What do you mean by collapsing? GPUTreeshap runs only on device memory, so I'm not sure how we can avoid host-to-device copies

File: src/model.h (hunk 11 above)
Comment (divyegala) [reply src/model.h:153]: How do we avoid error checking by using `->` instead of `.`?

Comment (wphicks) [reply src/model.h:153]: Edited: Had a typo in my suggestion.

File: src/model.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (divyegala) [reply src/model.h:133]: This method works for all buffers, but none of them can be const-reference for various reasons.

File: src/gpu_treeshap_model.h [hunk 16]
```diff
@@ -35,9 +35,11 @@ struct TreeShapModel<rapids::DeviceMemory> {
   TreeShapModel(
       device_id_t device_id, cudaStream_t stream,
       std::shared_ptr<TreeliteModel> tl_model)
-      : device_id_{device_id}, raft_handle_{stream}, tl_model_{tl_model}
+      : device_id_{device_id}, raft_handle_{stream}, tl_model_{tl_model},
+        path_info_{[this]() {
+          return ML::Explainer::extract_path_info(tl_model_->handle());
+        }()}
```
Comment (wphicks) [inline src/gpu_treeshap_model.h:41]:
```suggestion
        path_info_{ML::Explainer::extract_path_info(tl_model_->handle()}
```

**Review by wphicks — CHANGES_REQUESTED** [review]

**Review by wphicks — APPROVED** [review]
LGTM!

**Pull request merged**